File: src/utils/incompleteReturnDetection.ts
import type { Item } from '../data/itemsService';
import type { Transaction } from '../data/transactionsService';
import type { ItemLineageEdge } from '../data/lineageEdgesService';

/**
 * An item has an incomplete return when ALL of the following are true:
 * 1. item.status === 'returned'
 * 2. The transaction is not a return transaction
 * 3. No lineage edge with movementKind 'returned' exists for this item from this transaction
 */
export function isIncompleteReturn(
  item: Pick<Item, 'id' | 'status' | 'transactionId'>,
  transaction: Pick<Transaction, 'transactionType'> | null,
  edgesFromTransaction: Pick<ItemLineageEdge, 'itemId' | 'movementKind'>[],
): boolean {
  if (item.status !== 'returned') return false;
  if (!transaction) return false;

  const txType = transaction.transactionType?.trim().toLowerCase();
  if (txType === 'return') return false;

  const hasReturnedEdge = edgesFromTransaction.some(
    (edge) => edge.itemId === item.id && edge.movementKind === 'returned',
  );
  if (hasReturnedEdge) return false;

  return true;
}

/**
 * Filter active items to find those with incomplete returns.
 * Returns array of item IDs.
 */
export function findIncompleteReturns(
  activeItems: Pick<Item, 'id' | 'status' | 'transactionId'>[],
  transaction: Pick<Transaction, 'transactionType'> | null,
  edgesFromTransaction: Pick<ItemLineageEdge, 'itemId' | 'movementKind'>[],
): string[] {
  return activeItems
    .filter((item) => isIncompleteReturn(item, transaction, edgesFromTransaction))
    .map((item) => item.id);
}

File: LedgeriOS/LedgeriOS/Logic/IncompleteReturnDetection.swift
/// Detects items whose status is "returned" but haven't been properly processed
/// through the return flow (no "returned" lineage edge from their current transaction).
enum IncompleteReturnDetection {

    /// An item has an incomplete return when ALL of the following are true:
    /// 1. item.status == "returned"
    /// 2. The transaction is not a Return transaction
    /// 3. No lineage edge with movementKind "returned" exists for this item
    static func isIncompleteReturn(
        itemId: String,
        itemStatus: String?,
        transactionType: String?,
        edgesFromTransaction: [LineageEdge]
    ) -> Bool {
        guard itemStatus == "returned" else { return false }

        // If no transaction context, can't determine completeness
        guard let txType = transactionType else { return false }

        // Item is already in a return transaction — return is complete
        if txType.trimmingCharacters(in: .whitespaces).lowercased() == "return" {
            return false
        }

        // Check if a "returned" edge exists for this item
        let hasReturnedEdge = edgesFromTransaction.contains {
            $0.itemId == itemId && $0.movementKind == "returned"
        }

        return !hasReturnedEdge
    }

    /// Filters items to find those with incomplete returns. Returns their IDs.
    static func findIncompleteReturns(
        items: [(id: String, status: String?)],
        transactionType: String?,
        edgesFromTransaction: [LineageEdge]
    ) -> [String] {
        items.compactMap { item in
            isIncompleteReturn(
                itemId: item.id,
                itemStatus: item.status,
                transactionType: transactionType,
                edgesFromTransaction: edgesFromTransaction
            ) ? item.id : nil
        }
    }
}
